named inference rule for tensor.select

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21795 Implement tensor.select(Dimname,int)
* #21789 Add at::dimname_to_position helper.
* **#21752 named inference rule for tensor.select**
* #21781 Disallow creation of tensors with duplicate names
* #21803 Rename Dimname::name to Dimname::full_name
* #21735 Copy NamedTensorMeta in TensorImpl::copy_tensor_data()
* #21648 Expose torch.empty(sizes, *, names, ...) to Python

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

Differential Revision: [D15833453](https://our.internmc.facebook.com/intern/diff/D15833453)

zou3519 committed:
named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

nairbv commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -458,7 +470,13 @@ Tensor select(const Tensor& self, int64_t dim, int64_t index) {
   sizes.erase(sizes.begin() + dim);
   strides.erase(strides.begin() + dim);
-  return self.as_strided(sizes, strides, storage_offset);
+  auto result = self.as_strided(sizes, strides, storage_offset);
```
It seems like as_strided should accept names, and then the rest of the code would just look like the existing code.
Will the reduction_op function be used for something else that it makes sense to break it out?

zou3519 replied:
> It seems like as_strided should accept names, and then the rest of the code would just look like the existing code.

That's an interesting proposal. `as_strided` can do pretty crazy things to tensors (change their sizes and strides). In a named-tensor-world, the user shouldn't use `as_strided` directly because it shouldn't be necessary. Because of that, I don't want to add a `names` argument to `as_strided`.

> Will the reduction_op function be used for something else that it makes sense to break it out?

It'll be useful for other functions that take an `int dim`.

nairbv commented on aten/src/ATen/NamedTensorUtils.h:
```diff
@@ -10,6 +11,11 @@ inline bool has_names(TensorList tensors) {
-}
+namespace namedinference {
+
+optional<std::vector<Dimname>> reduction_op(optional<DimnameList> self_names, int64_t dim);
```
Is there a better name we could use for this function? As a helper function it's not what I'd usually think of as being reduction operator.

zou3519 replied:
Maybe something like `namedinference::remove_dimension`

nairbv replied:
yeah that would be fine. reduce_op to mean just feels like it implies an actual tensor reduction operator like sum or mean

zou3519 replied:
Done. I named it `namedinference::erase_name`.

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head

zou3519 committed:
Update on "named inference rule for tensor.select"

named inference rule for tensor.select

`x.select(dim, idx)` returns a tensors with names equal to x.names
without x.names[idx].

In addition, `x.as_strided(...)` returns a tensor with no names because
it loses naming information.

Finally, I modified the named tensor tests to save some keystrokes.
`skipIfNamedTensorDisabled` is applied to all `test_*` methods of
TestNamedTensor.

Test Plan:
- `python test/test_namedtensor.py` [namedtensor ci]

gh-metadata: pytorch pytorch 21752 gh/zou3519/50/head
